apps/wpcom-block-editor/src/common/tracking.js

```diff
@@ -0,0 +1,133 @@
+					`properties are not supported by Tracks. Check '${ key }' on`;
+				console.error( errorMessage, eventProperties ); //eslint-disable-line no-console
+				return;
+			}
+
+			if ( ! /^[a-z_][a-z0-9_]*$/.test( key ) ) {
+				//eslint-disable-next-line no-console
+				console.error(
+					'Tracks: Event `%s` will be rejected because property name `%s` does not match /^[a-z_][a-z0-9_]*$/. ' +
+						'Please use a compliant property name.',
+					eventName,
+					key
+				);
+			}
+		}
+	}
+
+	// Remove properties that have an undefined value
+	// This allows a caller to easily remove properties from the recorded set by setting them to undefined
+	eventProperties = omit( eventProperties, isUndefined );
+
+	eventProperties = {
+		...eventProperties,
+		blog_id: blogId,
+	};
+
+	tracksDebug( 'Recording event "%s" with actual props %o', eventName, eventProperties );
+
+	// window._tkq.push( [ 'identifyUser', userId, userUsername ] );
+	window._tkq.push( [ 'recordEvent', eventName, eventProperties ] );
+};
+
+// looks up the block name based on its id
+const getTypeForBlockId = blockId => {
+	const block = window.wp.data.select( 'core/block-editor' ).getBlock( blockId );
+	return block ? block.name : null;
+};
+
+// A lot of actions accept either string (block id) or an array of multiple string
+// when operating with multiple blocks selected.
+// This is a convenience method that processes the first argument of the action (blocks)
+// and calls your tracking for each of the blocks involved in the action.
+const getBlocksTracker = eventName => blocks => {
+	if ( typeof blocks === 'string' ) {
+		// single block
+		tracksDebug( 'TRACK:', eventName, { block_name: getTypeForBlockId( blocks ) } );
+	} else {
+		// multiple blocks at the same time
+		const blockNames = blocks.map( getTypeForBlockId );
+		// track separately for each block
+		blockNames.forEach( block_name => {
+			tracksDebug( 'TRACK:', eventName, { block_name } );
+		} );
+	}
+};
+
+// tracker can be
+// - string - which means it is an event name and should be tracked as such automatically
+// - function - in case you need to load additional properties from the action.
+//              you are fully responsible to call the tracking
+const REDUX_TRACKING = {
+	'core/editor': {
+		undo: 'wpcom_block_editor_undo_performed',
+		redo: 'wpcom_block_editor_redo_performed',
+	},
+	'core/block-editor': {
+		moveBlocksUp: getBlocksTracker( 'wpcom_block_moved_up' ),
+		moveBlocksDown: getBlocksTracker( 'wpcom_block_moved_down' ),
+		removeBlocks: getBlocksTracker( 'wpcom_block_deleted' ),
+		removeBlock: getBlocksTracker( 'wpcom_block_deleted' ),
+		moveBlockToPosition: getBlocksTracker( 'wpcom_block_moved_via_dragging' ),
+	},
+};
+
+// intercept dispatch function and add tracking for actions that need it
+window.wp.data.use( registry => ( {
+	dispatch: namespace => {
+		const actions = { ...registry.dispatch( namespace ) };
```

Comment:
Is the idea behind the `{ ... }` here to copy the object?

If yes, I think it might make sense to follow the functional programming credo of not mutating objects here, but always creating a new one -- i.e. rather than modifying `actions[ actionName ] = ...` in a `forEach` loop, you could create a new object via `.reduce()`.

Something like

```js
Object.keys( trackers ).reduce( ( newActions, actionName ) => {
	/*...*/
	return { ...newActions, actionName: ( ...args ) => {
		/*...*/
	}
}, {} ) // the empty object here is the initial value for the reduce
```

(You might even be able to use `.map()`, or lodash's `mapValues()`. Haven't thought that through tho.)

I find these a bit clearer to read, as they state intent more than a generic `forEach` does, but I'll admit that it might be partly because I've got more used to that style now.